From f218ddf248a92efe705b9fe00e0a1f32eaefb544 Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Tue, 4 Nov 2014 22:07:17 -0500
Subject: [powerpc] numa: ensure per-cpu NUMA mappings are correct on topology update

Message-id: <1415138837-28185-9-git-send-email-gduarte@redhat.com>
Patchwork-id: 99392
O-Subject: [RHEL7.1 PATCH BZ 1150097 8/8] powerpc/numa: ensure per-cpu NUMA mappings are correct on topology update
Bugzilla: 1150097
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

BZ 1150097
Upstream Status: https://git.kernel.org/cgit/linux/kernel/git/mpe/linux.git/commit/?h=fixes&id=2c0a33f9861d38631245f7ef434ecad3413324fb

Author: Nishanth Aravamudan <nacc@linux.vnet.ibm.com>
Date: Fri, 17 Oct 2014 17:50:40 -0700

    powerpc/numa: ensure per-cpu NUMA mappings are correct on topology update

    We received a report of warning in kernel/sched/core.c where the sched
    group was NULL on an LPAR after a topology update. This seems to occur
    because after the topology update has moved the CPUs, cpu_to_node is
    returning the old value still, which ends up breaking the consistency of
    the NUMA topology in the per-cpu maps. Ensure that we update the per-cpu
    fields when we re-map CPUs.

    Signed-off-by: Nishanth Aravamudan <nacc@linux.vnet.ibm.com>
    Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/mm/numa.c b/arch/powerpc/mm/numa.c
index 2e30700..141a8a8 100644
--- a/arch/powerpc/mm/numa.c
+++ b/arch/powerpc/mm/numa.c
@@ -1515,11 +1515,14 @@ static int update_cpu_topology(void *data)
  cpu = smp_processor_id();
 
  for (update = data; update; update = update->next) {
+  int new_nid = update->new_nid;
   if (cpu != update->cpu)
    continue;
 
   unmap_cpu_from_node(cpu);
-  map_cpu_to_node(cpu, update->new_nid);
+  map_cpu_to_node(cpu, new_nid);
+  set_cpu_numa_node(cpu, new_nid);
+  set_cpu_numa_mem(cpu, local_memory_node(new_nid));
   vdso_getcpu_init();
  }
 
-- 
1.7.1